addrs/provider_config.go

```diff
@@ -279,9 +279,14 @@ func ParseLegacyAbsProviderConfig(traversal hcl.Traversal) (AbsProviderConfig, t
 		return ret, diags
 	}
 
-	// We always assume legacy-style providers in legacy state
+	// We always assume legacy-style providers in legacy state ...
 	if tt, ok := remain[1].(hcl.TraverseAttr); ok {
-		ret.Provider = NewLegacyProvider(tt.Name)
+		// ... unless it's the builtin "terraform" provider, a special case.
+		if tt.Name == "terraform" {
```

Comment:
Is there a way to lean on the logic in `ImpliedProviderForUnqualifiedType` here? https://github.com/hashicorp/terraform/blob/2dd64a7816598d08fb06c17d27cbfb5b8c0e81da/addrs/provider.go#L105-L115

I'm pretty sure it practically doesn't matter (are we ever going to add more implied-builtin providers?), but having similar lookup logic in two places feels a little error-prone.

Reply:
Even if we were to add a new builtin provider later, I don't think we'd need this codepath to support it because there would be no reason for a pre-0.13 state snapshot to ever include an unqualified reference to it.

I do share your sense of "ugh" at putting what is more-or-less the same logic in two places, but in this case I think (counter-intuitively) it's the better strategy because we _don't_ want this to get updated if we add new builtin providers in future: what's accepted for upgrading here will be fixed in place forever after our final release, bug fixes notwithstanding.